src/core/varinfo.jl

```diff
@@ -33,138 +33,240 @@ Base.convert(::Type{Tuple}, vn::VarName) = uid(vn)
+Base.haskey(vi::VarInfo, vn::VarName) = haskey(vi.idcs, uid(vn))
+
+# Get all indices of variables belonging to gid or 0
+groupidcs(vi::VarInfo, gid::Int, spl=nothing) = begin
+  if spl == nothing || isempty(spl.alg.space)
+    filter(i -> vi.gids[i] == gid || vi.gids[i] == 0, 1:length(vi.gids))
+  else
+    filter(i -> (vi.gids[i] == gid || vi.gids[i] == 0) && (vi.uids[i][2] in spl.alg.space), 1:length(vi.gids))
+  end
+end
+
+# Get all values of variables belonging to gid or 0
+groupvals(vi::VarInfo, gid::Int, spl=nothing) = map(i -> vi.vals[i], groupidcs(vi, gid, spl))
+
+# Get all uids of variables belonging to gid or 0
+groupuids(vi::VarInfo, gid::Int, spl=nothing) = map(i -> vi.uids[i], groupidcs(vi, gid, spl))
+
+retain(vi::VarInfo, gid::Int, n_retain) = begin
+  # NOTE: the sanity check below is commented because Void
+  #       and standalone samplers uses gid = 0
+  # @assert ~(gid == 0) "[retain] wrong use of retain: gid = 0"
+
+  # Get all indices of variables belonging to gid
+  gidcs = filter(i -> vi.gids[i] == gid, 1:length(vi.gids))
+  l = length(gidcs)
+
+  # Remove corresponding entries
+  for i = l:-1:(n_retain + 1)
+    delete!(vi.idcs, vi.uids[gidcs[i]])
+    splice!(vi.uids, gidcs[i])
+    splice!(vi.vals, gidcs[i])
+    splice!(vi.dists, gidcs[i])
+    splice!(vi.gids, gidcs[i])
+    splice!(vi.istrans, gidcs[i])
+  end
+
+  # Rebuild index dictionary
+  for i = 1:length(vi.uids)
+    vi.idcs[vi.uids[i]] = i
+  end
+
+  vi
+end
+
+# Add a new entry to VarInfo
+addvar!(vi::VarInfo, vn::VarName, val, dist::Distribution, gid=0, istrans=false) = begin
+  @assert ~(uid(vn) in uids(vi)) "[addvar!] attempt to add an exisitng variable $(sym(vn)) ($(uid(vn))) to VarInfo (keys=$(keys(vi))) with dist=$dist, gid=$gid, istrans=$istrans"
   vi.idcs[uid(vn)] = length(vi.idcs) + 1
+  push!(vi.uids, uid(vn))
   push!(vi.vals, val)
-  push!(vi.syms, vn.sym)
   push!(vi.dists, dist)
+  push!(vi.gids, gid)
+  push!(vi.istrans, istrans)
 end
 
-syms(vi::VarInfo) = union(Set(vi.tsyms), Set(vi.syms))
-uids(vi::VarInfo) = union(Set(keys(vi.idcs)), Set(vi.names))
-
-# TODO: change below after randr() is unified
-Base.keys(vi::VarInfo) = map(t -> VarName(t...), keys(vi.idcs))
-Base.haskey(vi::VarInfo, vn::VarName) = haskey(vi.idcs, uid(vn))
-
+# This method is use to generate a new VarName with the right count
 nextvn(vi::VarInfo, csym::Symbol, sym::Symbol, indexing::String) = begin
-  # TODO: update this method when VarInfo internal structure is updated
+  # TODO: update this method when implementing the sanity check
   VarName(csym, sym, indexing, 1)
 end
 
-# TODO: below should be updated when the field group is add to InferenceAlgorithm
-rand(vi::VarInfo, vn::VarName, dist::Distribution, method::Symbol) = begin
-  if method == :byname
-    randrn(vi, vn, dist)
-  elseif method == :bycounter
-    randrc(vi, vn, dist)
+# Main behaviour control of rand() depending on sampler type and if sampler inside
```

Comment:
`inside` controls whether `rand()` is dealing with variables within the space of sampler which calls `rand()` or not.